src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,286 @@
+
+        if (replaceString.indexOf(searchString) === -1) {
+            // normal case: replaceString doesn't contain searchString so we
+            // can decrement searchResultCount because we just replaced one 
+            // occurence of searchString
+            searchResultCount --;
+        } else {
+            // edge case: replaceString contains one or more instances of
+            // searchString which means the searchResultCount will either 
+            // stay the same if the number of instances is one or increase
+            // if it's more than one
+            var matches = replaceString.match(regex);
+            if (matches) {
+                searchResultCount += matches.length - 1;
+            }
+        }
+
+        if (searchIndex >= searchResultCount) {
+            searchIndex = searchResultCount - 1;
+        }
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ question, hints, json: particle, searchIndex });
+    },
+
+    render() {
+        var displayCount = this.state.searchResultCount;
+        var disabled = displayCount === 0;
+
+        var displayIndex = this.props.searchIndex + 1;
+        if (displayIndex > displayCount) {
+            displayIndex = displayCount;
+        }
+
+        var gridSpace = 8;
+
+        var style = {
+            padding: 10,
+            position: 'fixed',
+            right: 0,
+            top: 0,
+            width: 300,
+            backgroundColor: '#EEE',
+            border: 'solid 1px #DDD',
+            zIndex: 100,
+            display: 'flex',
+            flexDirection: 'row'
+        };
+
+        var labelStyle = {
+            display: 'inline-block',
+            textAlign: 'right',
+            width: 60,
+            marginRight: gridSpace
+        };
+
+        var inputStyle = {
+            display: 'inline-block',
+            width: '100%',
+            boxSizing: 'border-box'
+        };
+
+        return <div style={style}>
+            <div style={{ flexShrink: 0, flexGrow: 0 }}>
+                <label style={labelStyle}>Search:</label>
+                <br />
+                <label style={labelStyle}>Replace:</label>
+                <br />
+                <span style={labelStyle}>{displayIndex} of {displayCount}</span>
+            </div>
+            <div style={{ flexShrink: 0, flexGrow: 1 }}>
+                <input
+                    ref="searchInput"
+                    type="text"
+                    value={this.props.searchString}
+                    onChange={this.updateSearchString}
+                    style={inputStyle} />
+                <br />
+                <input
+                    ref="replaceInput"
+                    type="text"
+                    value={this.props.replaceString}
+                    onChange={this.updateReplaceString}
+                    style={inputStyle} />
+                <br />
+                <div>
+                    <button
```

Comment:
we probably want `type="button"` on each of these. the default button type is `"submit"`, which is rarely what we want (and means that when embedded in another page, it can cause weird actions)